Add docs for MessagePack (#21090)
dotnet/AspNetCore.Docs: aspnetcore/signalr/messagepackhubprotocol.md

> [!NOTE]
> At this time, there are no configuration options for the MessagePack protocol on the JavaScript client.

> ### Java client

Author: Does anyone know why this doc repeats itself 3 times?

Reviewer: Yeah, different versions. Each one is between a `::: moniker range=">= aspnetcore-5.0"` and `::: moniker-end` and has different content based on which version you're referencing

Author: Ah, didn't notice that. I'll remove the new stuff from the pre-5.0 sections

> To enable MessagePack in the Java, first install the `com.microsoft.signalr.messagepack` package. If using Gradle, add the following line to the `dependencies` section of your *build.gradle* file:

Reviewer 2: Something like:

```suggestion
To enable MessagePack with Java, install the `com.microsoft.signalr.messagepack` package. When using Gradle, add the following line to the `dependencies` section of the *build.gradle* file:
```

```gradle
implementation 'com.microsoft.signalr.messagepack:signalr-messagepack:5.0.0'
```

> If using Maven, add the following lines inside the `<dependencies>` element of your *pom.xml* file:

Reviewer 2: ```suggestion
When using Maven, add the following lines inside the `<dependencies>` element of the *pom.xml* file:
```

> [!code-xml[pom.xml dependency element messagePack](java-client/sample/pom.xml?name=snippet_dependencyElement_messagePack)]

Author: This didn't work, anybody know what the right syntax is?

Reviewer 3: Looks like it worked to me: https://review.docs.microsoft.com/en-us/aspnet/core/signalr/messagepackhubprotocol?view=aspnetcore-5.0&branch=pr-en-us-21090#java-client

> Then call `withHubProtocol(new MessagePackHubProtocol())` on `HubConnectionBuilder`.

Reviewer 2: ```suggestion
Call `withHubProtocol(new MessagePackHubProtocol())` on `HubConnectionBuilder`.
```

For more information on this limitation, see GitHub issue [aspnet/SignalR#2937](https://github.com/aspnet/SignalR/issues/2937).

> ### Passing Class information in Java

Reviewer: This isn't specific to MessagePack is it?

Author: No, but it's only required in MessagePack, JSON still works without it. I could move this to java-client.md and call out the differences there.

Reviewer 3: How does JSON manage to work without it given the type erasure? It's also probably worth pointing out type erasure is the reason a TypeReference is necessary.

Author: Oh, you're right, I was remembering wrong - JSON had the same issue with more complex types but we hadn't noticed before. I'll move this segment & expand it now.

Author: https://github.com/dotnet/AspNetCore.Docs/pull/21090/commits/fd51aba3760da7845281f792159238d121ee3741

> When calling the `on()`, `invoke()`, or `stream()` methods of `HubConnection` in the Java client, MessagePack users should pass a `Type` object rather than a `Class<?` object to describe the `Object`(s) they are passing to the method. A `Type` can be acquired using the provided `TypeReference` class. For example, if you have a custom class named `Foo` and another named `Bar`, you can get their `Type`s with the following code:

Reviewer 2: ```suggestion
When calling the `on`, `invoke`, or `stream` methods of `HubConnection` in the Java client, MessagePack users should pass a `Type` object rather than a `Class<?` object to describe each `Object` passed to the method. A `Type` can be acquired using the provided `TypeReference` class. For example, using a custom classes named `Foo` and `Bar`, the following code gets each `Type`:
```

```java
Type fooType = new TypeReference<Foo>() { }).getType();
Type barType = new TypeReference<Bar>() { }).getType();
```

> Furthermore, when calling one of these methods with one or more object types, you should use the generics syntax when invoking the method. For example, if registering an `on` handler for a method named `func`, which takes as arguments a `Foo` object and a `Bar` object, you could use the following code to set an action to print the two objects:

Reviewer 2: ```suggestion
When calling one of these methods with one or more object types, use the generics syntax when invoking the method. For example, when registering an `on` handler for a method named `func`, which takes as arguments a `Foo` object and a `Bar` object, use the following code to set an action to print the two objects:
```

```java
hubConnection.<Foo, Bar>on("func", (param1, param2) ->{
    System.out.println(param1);
    System.out.println(param2);
}, fooType, barType);
```

> The exception to this rule is primitives such as `int`, for which you should use the built-in `int.class`.

Reviewer 2: ```suggestion
The exception to this rule is primitives such as `int`. When using `int`, use the built-in `int.class`.
```

### Chars and Strings in Java

> In the java client, `char` objects will be serialized as one-character `String`s. This is in contrast with the C# and JavaScript client, which serialize them as `short`s. The MessagePack spec itself does not define behavior for `char`s, so it is up to the library author to determine how to serialize them. The difference in behavior between our clients is a result of the libraries we used for our implementations.

Reviewer 2: The following doesn't localize
```
`String`s
`short`s
`char`s
```

So rewrite as strings, or series of one character string types, or something like that.
